### DISCUSSION TITLE
Support for multi-config based monorepo setup in ESLint 9

### DISCUSSION THREAD

**Discussion by fabis94** [eslint/eslint#18574] [Q&A]
I've been busy migrating a monorepo to ESLint 9, where each of the monorepo packages had its own eslint config. That is still the case in the new setup and everything works fine, except when I want to lint the entire monorepo with one command (e.g. for linting all staged files).

ESLint <=8 picked the most appropriate config file for each linted file, but with ESLint 9 that no longer seems the case, and it seems to expect a single config file.

I definitely do not want to do that - the config for each monorepo package should be its own responsibility, and I should not have to install all package linting deps at the root level just so I can have a single config. Not only that, but each package might have a different version of the same linting dep (e.g. different versions of an eslint plugin). 

How would you suggest dealing with this problem? I've been trying to create a special eslint config that imports all of the other ones and merges them together, but that hasn't been working very well. I'm not aware of any special merging utilities, and the manual low-level merging I've been doing doesn't work well because of having to adjust all of the glob patterns and mismatched eslint plugin versions (or them not being available at root level)

**Comment by nzakas** [comment]
The easiest thing to do right now is to have a config in each package directory, and then have a root config that reads in each of the package-specific configs. 

We are working on changing the lookup to work more like eslintrc:
https://github.com/eslint/eslint/issues/18385#issuecomment-2115516328

**Reply by fabis94** [reply]
@nzakas Yea, but that's exactly what doesn't work for me and I don't expect to work for any medium-to-large sized project. I've got a monorepo with like ~14 packages in there, with a couple of those being Nuxt apps which have a very complicated eslint config (see @nuxt/eslint).

Even if I adjust all the glob patterns when merging them all together, I'm still getting errors with plugins being missing or TS parserOptions being missing or if I resolve those through hacks its something else. All of these configs work fine individually, it's only when merging them all into one does it all break down.

So instead what I did was just build a utility script that invokes ESLint in a child process separately for each package: https://github.com/specklesystems/speckle-server/blob/fabians/eslint-9/utils/eslint-projectwide.mjs

**Reply by nzakas** [reply]
> So instead what I did was just build a utility script that invokes ESLint in a child process separately for each package

That works too. :1:

**Comment by nzakas** [comment]
Update: We now have implemented an experimental config lookup algorithm that works more like the eslintrc-style lookup, allowing different directories to have different config files. Please try it out and let us know any feedback.

https://eslint.org/docs/latest/use/configure/configuration-files#experimental-configuration-file-resolution

**Reply by jclaveau** [reply]
And it's a true pleasure to use!
